File: app/src/main/java/com/dpis/module/quirks/WechatDpiSheetBinder.java
package com.dpis.module.quirks;

import com.dpis.module.DpisConfigStore;

import com.dpis.module.viewport.DpiConfig;

import com.dpis.module.DpisApplication;
import com.dpis.module.R;
import com.dpis.module.appconfig.WechatDpiConfig;

import com.dpis.module.ui.DialogWindowSizer;

import android.text.Editable;
import android.text.TextWatcher;
import android.view.HapticFeedbackConstants;
import android.view.View;

import com.google.android.material.color.MaterialColors;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

public final class WechatDpiSheetBinder {
    private WechatDpiSheetBinder() {
    }

    public static void bind(View dialogView, String packageName, Runnable onValidationChanged) {
        View row = row(dialogView);
        TextInputLayout inputLayout = inputLayout(dialogView);
        TextInputEditText inputView = inputView(dialogView);
        if (row == null || inputLayout == null || inputView == null) {
            return;
        }
        if (!WechatDpiConfig.appliesTo(packageName)) {
            row.setVisibility(View.GONE);
            return;
        }
        row.setVisibility(View.VISIBLE);
        View helpButton = helpButton(dialogView);
        if (helpButton != null) {
            helpButton.setOnClickListener(v -> {
                v.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY);
                showHelpDialog(v);
            });
        }
        DpisConfigStore store = DpisApplication.getActiveHookConfigStore(
                dialogView.getContext());
        Integer initial = store != null ? store.getWechatDpi(packageName) : null;
        inputView.setText(initial != null ? String.valueOf(initial) : "");
        inputView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                updateValidationState(inputLayout, inputView);
                if (onValidationChanged != null) {
                    onValidationChanged.run();
                }
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });
        updateValidationState(inputLayout, inputView);
    }

    public static boolean isInputValid(View dialogView) {
        TextInputEditText inputView = inputView(dialogView);
        if (inputView == null || inputView.getText() == null) {
            return true;
        }
        return isInputValid(inputView.getText().toString());
    }

    /** Validates the retained editor value without requiring a legacy View. */
    public static boolean isInputValid(String rawValue) {
        return WechatDpiConfig.isInputValid(rawValue);
    }

    public static void bindDoneAction(android.widget.TextView.OnEditorActionListener listener,
            View dialogView) {
        TextInputEditText inputView = inputView(dialogView);
        if (inputView != null) {
            inputView.setOnEditorActionListener(listener);
        }
    }

    public static TextInputEditText inputViewForFocus(View dialogView) {
        return inputView(dialogView);
    }

    public static boolean save(View dialogView, String packageName, boolean dpisEnabled,
            DpisConfigStore store) {
        TextInputEditText inputView = inputView(dialogView);
        String rawValue = inputView != null && inputView.getText() != null
                ? inputView.getText().toString()
                : null;
        return save(rawValue, packageName, dpisEnabled, store);
    }

    /**
     * Saves the raw editor value shared by Compose and legacy editor surfaces.
     */
    public static boolean save(String rawValue, String packageName, boolean dpisEnabled,
            DpisConfigStore store) {
        if (!WechatDpiConfig.appliesTo(packageName)) {
            return true;
        }
        if (store == null || !isInputValid(rawValue)) {
            return false;
        }
        Integer dpi = WechatDpiConfig.parseOrNull(rawValue);
        boolean saved = store.setWechatDpi(packageName, dpi);
        if (saved) {
            WechatDpiPropertySyncer.publishDpiAsync(packageName, dpisEnabled ? dpi : null);
        }
        return saved;
    }

    public static void publishForDpisState(String packageName, boolean dpisEnabled) {
        if (!WechatDpiConfig.appliesTo(packageName)) {
            return;
        }
        DpisConfigStore store = DpisApplication.getActiveHookConfigStore(null);
        Integer dpi = store != null && dpisEnabled ? store.getWechatDpi(packageName) : null;
        WechatDpiPropertySyncer.publishDpiAsync(packageName, dpi);
    }

    public static void clearDraft(View dialogView) {
        TextInputEditText inputView = inputView(dialogView);
        if (inputView != null) {
            inputView.setText("");
        }
    }

    public static String captureDraft(View dialogView) {
        TextInputEditText inputView = inputView(dialogView);
        if (inputView == null || inputView.getText() == null) {
            return null;
        }
        return inputView.getText().toString();
    }

    public static void applyDraft(View dialogView, String rawValue) {
        if (rawValue == null) {
            return;
        }
        TextInputEditText inputView = inputView(dialogView);
        TextInputLayout inputLayout = inputLayout(dialogView);
        if (inputView == null) {
            return;
        }
        inputView.setText(rawValue);
        updateValidationState(inputLayout, inputView);
    }

    private static Integer readDpiOrNull(View dialogView) {
        TextInputEditText inputView = inputView(dialogView);
        if (inputView == null || inputView.getText() == null) {
            return null;
        }
        return WechatDpiConfig.parseOrNull(inputView.getText().toString());
    }

    private static void updateValidationState(TextInputLayout inputLayout,
            TextInputEditText inputView) {
        if (inputLayout == null || inputView == null) {
            return;
        }
        String raw = inputView.getText() != null ? inputView.getText().toString() : "";
        boolean valid = WechatDpiConfig.isInputValid(raw);
        int defaultStrokeColor = MaterialColors.getColor(
                inputLayout, com.google.android.material.R.attr.colorOutline);
        int errorStrokeColor = MaterialColors.getColor(
                inputLayout, androidx.appcompat.R.attr.colorError);
        inputLayout.setError(null);
        inputLayout.setErrorEnabled(false);
        inputLayout.setBoxStrokeColor(valid ? defaultStrokeColor : errorStrokeColor);
    }

    private static View row(View dialogView) {
        return dialogView.findViewById(R.id.dialog_wechat_dpi_row);
    }

    private static TextInputLayout inputLayout(View dialogView) {
        return dialogView.findViewById(R.id.dialog_wechat_dpi_input_layout);
    }

    private static TextInputEditText inputView(View dialogView) {
        return dialogView.findViewById(R.id.dialog_wechat_dpi_input);
    }

    private static View helpButton(View dialogView) {
        return dialogView.findViewById(R.id.dialog_wechat_dpi_help_button);
    }

    private static void showHelpDialog(View anchor) {
        if (anchor == null) {
            return;
        }
        androidx.appcompat.app.AlertDialog dialog = new MaterialAlertDialogBuilder(anchor.getContext())
                .setTitle(R.string.dialog_wechat_dpi_help_title)
                .setMessage(R.string.dialog_wechat_dpi_help_message)
                .setPositiveButton(R.string.dialog_close_button, null)
                .create();
        dialog.setOnShowListener(d -> DialogWindowSizer.applyStandardWidth(
                dialog, anchor.getContext()));
        dialog.show();
    }
}

File: app/src/main/java/com/dpis/module/quirks/presentation/WechatDpiSheetBinder.kt
package com.dpis.module.quirks.presentation

import android.text.Editable
import android.text.TextWatcher
import android.view.HapticFeedbackConstants
import android.view.View
import android.widget.TextView
import com.dpis.module.DpisApplication
import com.dpis.module.R
import com.dpis.module.appconfig.WechatDpiConfig
import com.dpis.module.config.DpisConfigStore
import com.dpis.module.quirks.WechatDpiEditor
import com.dpis.module.ui.DialogWindowSizer
import com.google.android.material.color.MaterialColors
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout

/** View adapter for the WeChat DPI editor row. Persistence lives in [WechatDpiEditor]. */
object WechatDpiSheetBinder {
    @JvmStatic
    fun bind(dialogView: View?, packageName: String?, onValidationChanged: Runnable?) {
        if (dialogView == null) return
        val row = row(dialogView)
        val inputLayout = inputLayout(dialogView)
        val inputView = inputView(dialogView)
        if (row == null || inputLayout == null || inputView == null) return
        if (!WechatDpiConfig.appliesTo(packageName)) {
            row.visibility = View.GONE
            return
        }
        row.visibility = View.VISIBLE
        helpButton(dialogView)?.setOnClickListener { view ->
            view.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY)
            showHelpDialog(view)
        }
        val store = DpisApplication.getActiveHookConfigStore(dialogView.context)
        val initial = store?.getWechatDpi(packageName)
        inputView.setText(if (initial != null) initial.toString() else "")
        inputView.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = Unit
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                updateValidationState(inputLayout, inputView)
                onValidationChanged?.run()
            }
            override fun afterTextChanged(s: Editable?) = Unit
        })
        updateValidationState(inputLayout, inputView)
    }

    @JvmStatic
    fun isInputValid(dialogView: View?): Boolean {
        val inputView = inputView(dialogView) ?: return true
        val text = inputView.text ?: return true
        return WechatDpiEditor.isInputValid(text.toString())
    }

    @JvmStatic
    fun bindDoneAction(listener: TextView.OnEditorActionListener?, dialogView: View?) {
        inputView(dialogView)?.setOnEditorActionListener(listener)
    }

    @JvmStatic
    fun inputViewForFocus(dialogView: View?): TextInputEditText? = inputView(dialogView)

    @JvmStatic
    fun save(
        dialogView: View?,
        packageName: String?,
        dpisEnabled: Boolean,
        store: DpisConfigStore?,
    ): Boolean {
        val inputView = inputView(dialogView)
        val rawValue = inputView?.text?.toString()
        return WechatDpiEditor.save(rawValue, packageName, dpisEnabled, store)
    }

    @JvmStatic
    fun clearDraft(dialogView: View?) {
        inputView(dialogView)?.setText("")
    }

    @JvmStatic
    fun captureDraft(dialogView: View?): String? {
        val inputView = inputView(dialogView) ?: return null
        return inputView.text?.toString()
    }

    @JvmStatic
    fun applyDraft(dialogView: View?, rawValue: String?) {
        if (rawValue == null) return
        val inputView = inputView(dialogView) ?: return
        inputView.setText(rawValue)
        updateValidationState(inputLayout(dialogView), inputView)
    }

    private fun updateValidationState(inputLayout: TextInputLayout?, inputView: TextInputEditText?) {
        if (inputLayout == null || inputView == null) return
        val raw = inputView.text?.toString().orEmpty()
        val valid = WechatDpiEditor.isInputValid(raw)
        val defaultStrokeColor = MaterialColors.getColor(
            inputLayout,
            com.google.android.material.R.attr.colorOutline,
        )
        val errorStrokeColor = MaterialColors.getColor(
            inputLayout,
            androidx.appcompat.R.attr.colorError,
        )
        inputLayout.error = null
        inputLayout.isErrorEnabled = false
        inputLayout.boxStrokeColor = if (valid) defaultStrokeColor else errorStrokeColor
    }

    private fun row(dialogView: View?) = dialogView?.findViewById<View>(R.id.dialog_wechat_dpi_row)
    private fun inputLayout(dialogView: View?) =
        dialogView?.findViewById<TextInputLayout>(R.id.dialog_wechat_dpi_input_layout)
    private fun inputView(dialogView: View?) =
        dialogView?.findViewById<TextInputEditText>(R.id.dialog_wechat_dpi_input)
    private fun helpButton(dialogView: View?) =
        dialogView?.findViewById<View>(R.id.dialog_wechat_dpi_help_button)

    private fun showHelpDialog(anchor: View?) {
        if (anchor == null) return
        val dialog = MaterialAlertDialogBuilder(anchor.context)
            .setTitle(R.string.dialog_wechat_dpi_help_title)
            .setMessage(R.string.dialog_wechat_dpi_help_message)
            .setPositiveButton(R.string.dialog_close_button, null)
            .create()
        dialog.setOnShowListener {
            DialogWindowSizer.applyStandardWidth(dialog, anchor.context)
        }
        dialog.show()
    }
}
